Improve: save files in a more reliable manner

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Use [QSaveFile](https://doc.qt.io/qt-5/qsavefile.html) for all cases where Mudlet is writing to disk (profiles, maps, notes, ...). QSaveFile does a few things under the hood to ensure a file can be written safely, and gives a nice error if no.
#### Motivation for adding to Mudlet
I'm hoping this'll address some cases of folks' settings resetting or some other weird things happening - which happens once in a blue moon, but still does. 

Doing disk IO on scale is something that is eventually bound to run into issues, I had the same with the Svo system. Hopefully this'll cover things off.
#### Other info (issues closed, discussion etc)
Fix https://github.com/Mudlet/Mudlet/issues/1342

vadi2 committed:
Migrate XMLExport

vadi2 committed:
Improve verbosity of XMLexport::saveXml

vadi2 committed:
Migrate ctelnet to QSaveFile

Remove .close() call as it's not necessary in this situation

vadi2 committed:
Revert "Migrate ctelnet to QSaveFile"

This reverts commit dd82a440bbdd4dbc6665fbafb5e2c71615945d13.

vadi2 committed:
Migrate connection dialog

vadi2 committed:
Update files

vadi2 committed:
Migrate more files

vadi2 committed:
Migrate the rest of the files

vadi2 commented:
List of modified functions for those testing:

dlgConnectionProfiles::extractSettingsFromProfile(pugi::xml_document& newPr
mudlet::overwriteAffixFile(QFile& aff, const QHash<QString, unsigned int>&
mudlet::overwriteDictionaryFile(QFile& dict, const QStringList& wl)
mudlet::saveDictionary(const QString& pathFileBaseName, QSet<QString>& word
mudlet::saveWindowLayout()
mudlet::scanDictionaryFile(QFile& dict, int& oldWC, QHash<QString, unsigned
XMLexport::saveXmlFile(QFile& file)
mudlet::prepareProfileDictionary(QString& hostName, QSet<QStrin
mudlet::prepareSharedDictionary()
mudlet::getDictionaryWordCount(QFile &dict)
mudlet::scanWordList(QStringList& wl, QHash<QString, unsigned int>& gc)
dlgConnectionProfiles::readProfileData(const QString& profile, const QSt
dlgIRC::readAppDefaultIrcNick()
Host::getPackageConfig(const QString& luaConfig, bool isModule)
mudlet::autodetectPreferredLanguage()
mudlet::readProfileData(const QString& profile, const QString& item)
TMainConsole::setCmdLineStyleSheet(QString& name,
TMap::writeJsonMapFile(QString& dest)
cTelnet::slot_replyFinished(QNetworkReply* reply)
dlgConnectionProfiles::saveProfileCopy(const QDir& newProfiledir, const pug
dlgConnectionProfiles::slot_setCustomColor()
dlgIRC::writeAppDefaultIrcNick(const QString& nick)
dlgNotepad::save()
dlgPackageExporter::writeConfigFile(const QString& stagingDirName, const QF
dlgTriggerEditor::slot_export()
mudlet::setEnableFullScreenMode(const bool state)
TConsole::closeEvent(QCloseEvent* event)
TConsole::slot_toggleReplayRecording()
TLuaInterpreter::handleHttpOK(QNetworkReply* reply)
TMainConsole::finalize()
TMainConsole::toggleLogging(bool isMessageEnabled)
TMap::slot_replyFinished(QNetworkReply* reply)
TMedia::slot_writeFile(QNetworkReply* reply)
TTextEdit::slot_toggleTimeStamps(const bool state)
Updater::recordUpdatedVersion() const
Updater::recordUpdateTime() const
Updater::slot_installOrRestartClicked(QAbstractButton* button, const QStrin
XMLexport::sanitizeForQxml(std::string& output)

vadi2 committed:
Fix to compile on Windows

SlySven commented on src/TConsole.cpp:
```diff
@@ -791,15 +791,17 @@ void TConsole::closeEvent(QCloseEvent* event)
                 // FIXME: https://github.com/Mudlet/Mudlet/issues/6316 - unchecked return value - we are not handling a failure to save the map!
                 mpHost->mpMap->serialize(out);
-                file_map.close();
+                if (!file.commit()) {
```
You can probably remove the `// FIXME :` here and in the next change - or do we need to put the error result on screen where (Windows users at least) might have a chance to see them?

vadi2 replied:
Windows users can see stdout on Windows fine, I put in a fix for that. The `serialize()`'s return status is still not handled so I think the fixme has to stay.

SlySven replied:
Okay...

SlySven commented on src/TConsole.cpp:
```diff
@@ -910,7 +914,9 @@ void TConsole::slot_toggleReplayRecording()
+        if (!mReplayFile.commit()) {
+            qDebug() << "TConsole::slot_toggleReplayRecording: error saving replay: " << mReplayFile.errorString();
+        }
         printSystemMessage(tr("Replay recording has been stopped. File: %1").arg(mReplayFile.fileName()) % QChar::LineFeed);
```
Given that the `(bool) QSaveFile::commit()` has returned false - should we not be telling the user in the next line that in fact their replay has been totally lost and that it *isn't* in the file specified as that doesn't exist...!

vadi2 replied:
Updated.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -588,7 +588,9 @@ void TLuaInterpreter::handleHttpOK(QNetworkReply* reply)
         }
 
-        localFile.close();
+        if (!localFile.commit()) {
```
Why are we not issuing a `nil` + error message for this?

vadi2 replied:
This is an event, not a function

SlySven replied:
Okay, I guess, but don't we have an event for failures as well - so shouldn't the execution be diverted to that instead on failure?

vadi2 replied:
We should, fixed.

SlySven commented on src/TMainConsole.cpp:
```diff
@@ -155,12 +155,15 @@ std::pair<bool, QString> TMainConsole::setCmdLineStyleSheet(const QString& name,
 {
-    QFile file(mudlet::getMudletPath(mudlet::profileDataItemPath, mpHost->getName(), qsl("autolog")));
+    const auto loggingPath = mudlet::getMudletPath(mudlet::profileDataItemPath, mpHost->getName(), qsl("autolog"));
+    QSaveFile file(loggingPath);
```
Actually this is once instance where using `QSaveFile` is pointless - as nothing is written to the file - instead it is used as a zero length *sentinel* file whose existence (or not) tells the profile when it starts that it is  or isn't to be logged from the start of the session...

:nerd_face:

vadi2 replied:
Can creating the file fail?

SlySven replied:
Well no more than it can with the conventional `QFile` - the important difference about the `QSaveFile` class is that it either completes and overwrites the entire contents of a previous save file instance or doesn't and leaves the prior instance untouched - but there are NO contents to be left in an inconsistent state by a partial/incomplete write in this situation so there is no benefit to using it...

vadi2 replied:
Sure, reverted.

vadi2 replied:
Actually no, I reverted the revert. We still can have a failure to open the file for writing for whatever reason.

SlySven replied:
Which the existing `QFile` class is perfectly capable of handling...

vadi2 replied:
Sure - but the current code didn't - and with qsavefile it does. It's a bit overkill, I agree.

SlySven commented on src/TMap.cpp:
```diff
@@ -2925,7 +2928,9 @@ std::pair<bool, QString> TMap::writeJsonMapFile(const QString& dest)
     }
     if (abort) {
-        file.close();
+        if (!file.commit()) {
```
This **isn't** an error - instead the user has clicked the cancel button on the progress dialogue and you should be invoking `QSaveFile::cancelWriting()` instead here.

vadi2 replied:
Definitely.

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -180,13 +180,16 @@ void TTextEdit::slot_toggleTimeStamps(const bool state)
         if (mpConsole->getType() == TConsole::MainConsole) {
-            QFile file(mudlet::getMudletPath(mudlet::profileDataItemPath, mpHost->getName(), qsl("autotimestamp")));
+            const auto filePath = mudlet::getMudletPath(mudlet::profileDataItemPath, mpHost->getName(), qsl("autotimestamp"));
+            QSaveFile file(filePath);
```
Again, this is a zero length *sentinel* file whose presence or absence determines behaviour (for the main console) on profile startup - so it doesn't help to use a `QSaveFile` for it...

SlySven commented on src/XMLexport.cpp:
```diff
@@ -305,34 +305,30 @@ void XMLexport::sanitizeForQxml(std::string& output)
-        }
+    bool success = saveXmlFile(file);
+    if (!success) {
+        printErrorMessage((file.error() != QFileDevice::NoError) ? file.errorString() : "XML document preparation failure");
```
Need to call `QSaveFile::cancelWriting()` here to abort and close the file - otherwise you are leaving it *open*...

vadi2 replied:
It should be OK:

> If [commit](https://doc.qt.io/qt-5/qsavefile.html#commit)() was not called and the QSaveFile instance is destroyed, the temporary file is discarded.

SlySven commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -722,22 +722,23 @@ QString dlgConnectionProfiles::readProfileData(const QString& profile, const QSt
         ofs << what;
-        file.close();
+        if (!file.commit()) {
+            qDebug() << "dlgConnectionProfiles::writeProfileData: error writing custom profile data: " << file.errorString();
```
It isn't "*custom*" data - it is any data that is stored as a file in the root of the profile's part of the file-system. You might actually want to say what the `item` is so someone who can actually see the debug output knows what has been messed up...
```suggestion
            qDebug() << "dlgConnectionProfiles::writeProfileData: error writing profile data: " << file.errorString();
```

vadi2 replied:
Updated.

SlySven commented on src/dlgPackageExporter.cpp:
```diff
@@ -829,7 +829,9 @@ void dlgPackageExporter::writeConfigFile(const QString& stagingDirName, const QF
 #endif
         out << mPackageConfig;
         out.flush();
```
Does `QSaveFile` have a `flush()` method - isn't it implied from the `commit()` one?

vadi2 replied:
It does, removed.

SlySven commented on src/mudlet.cpp:
```diff
@@ -3708,14 +3711,17 @@ void mudlet::setEnableFullScreenMode(const bool state)
         mEnableFullScreenMode = state;
-        QFile file_use_smallscreen(mudlet::getMudletPath(mudlet::mainDataItemPath, qsl("mudlet_option_use_smallscreen")));
+        auto filePath = mudlet::getMudletPath(mudlet::mainDataItemPath, qsl("mudlet_option_use_smallscreen"));
+        QSaveFile file(filePath);
```
A third zero-length *sentinel* file that doesn't need to be a `QSaveFile` IMHO.

SlySven commented on src/mudlet.cpp:
```diff
@@ -4168,19 +4179,18 @@ int mudlet::scanWordList(QStringList& wl, QHash<QString, unsigned int>& gc)
-    QString affixPathFileName(getMudletPath(mudlet::profileDataItemPath, hostName, qsl("profile.aff")));
-    QFile dictionary(dictionaryPathFileName);
-    QFile affix(affixPathFileName);
+    QString dictionaryPath(getMudletPath(mudlet::profileDataItemPath, hostName, qsl("profile.dic")));
```
:-1: I tend to use "pathFileName" because it tells me that the content is a **string** not a ***file*** and contains BOTH elements (directory AND file in that directory) - changing it to just "path" (or "pathName") loses that information...

vadi2 replied:
It leads to some very verbose and uncomfortable to work with code. You don't need to be reminded about the fact every single time you invoke the variable, so how about we just mention this fact at the point of definition?

SlySven commented:
Although you are generating more error messages - which is good - most users aren't going to see them - which is bad.

Are you considering a second PR to paint suitable error messages up in the main console?

vadi2 commented:
> most users aren't going to see them - which is bad.

That makes sense, but that assumes all of our users are technically experienced folks who'd know what the error messages are about, whenever they are relevant, and what to do with them! 

I think most people wouldn't know what to do with them, so I don't see not showing them in the main console as a bad thing. A fair bit of players would get confused, and some might even get overwhelmed when they're getting error messages that they have no idea about and have no idea what to do with.

> Are you considering a second PR to paint suitable error messages up in the main console?

No, not at the moment. stderr for now is good enough, until we get a dedicated logging window for the technically inclined folks who'd like to know more.

vadi2 committed:
Dont say recording saved if it failed

SlySven commented on src/TConsole.cpp:
```diff
@@ -916,8 +916,10 @@ void TConsole::slot_toggleReplayRecording()
     } else {
         if (!mReplayFile.commit()) {
             qDebug() << "TConsole::slot_toggleReplayRecording: error saving replay: " << mReplayFile.errorString();
+            printSystemMessage(tr("Replay recording has been stopped, but couldn't be saved.").arg(mReplayFile.fileName()) % QChar::LineFeed);
```
No positional parameter to take the `.arg(...)` Qt will moan about that at runtime (I know because I made the same mistake myself yesterday! :laughing:) : 
```suggestion
            printSystemMessage(tr("Replay recording has been stopped, but couldn't be saved.") % QChar::LineFeed);
```

vadi2 replied:
Thanks, fixed.

vadi2 committed:
Review feedback:

Fixed error message in TConsole::slot_toggleReplayRecording. Fixed error handling in TLuaInterpreter::handleHttpOK. Fixed error handling in TMainConsole::toggleLogging. Fixed error handling in TMap::writeJsonMapFile. Fixed error message in dlgConnectionProfiles::writeProfileData. Prepared profile dictionary in mudlet::prepareProfileDictionary.

vadi2 commented:
@SlySven have a look at reviewing again.

vadi2 committed:
Fix compiling

SlySven commented on src/TConsole.cpp:
```diff
@@ -832,15 +834,17 @@ void TConsole::closeEvent(QCloseEvent* event)
                     // FIXME: https://github.com/Mudlet/Mudlet/issues/6316 - unchecked return value - we are not handling a failure to save the map!
                     mpHost->mpMap->serialize(out);
-                    file_map.close();
+                    if (!file.commit()) {
```
Can the `TMap::serialize(...)` call fail (and return `false`) yet the `QSaveFile::commit()` succeed? - should we not test BOTH in some way?

vadi2 replied:
Yes it can. I am only looking to use qsavefile to save files more reliably in this PR.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -568,7 +568,9 @@ void TLuaInterpreter::handleHttpOK(QNetworkReply* reply)
         }
 
-        localFile.flush();
+        if (!localFile.commit()) {
```
:thinking: Presumably - if the `QSaveFile::commit()` has failed then automatically the `QSaveFile::error()` value will NOT be **`QFileDevice::NoError`** (and I guess `QSaveFile::NoError` == `QFile::NoError` == `QFileDevice::NoError`)?

vadi2 replied:
Right.

SlySven commented on src/TMainConsole.cpp:
```diff
@@ -1213,34 +1214,36 @@ void TMainConsole::finalize()
-    return false;
+    bool saved = mpHost->mpMap->serialize(out, saveVersion);
+    if (saved && !file_map.commit()) {
+        qDebug() << "Error saving map: " << file_map.errorString();
```
If `saved` is `false` there isn't going to be an error message is there? Unless it was generated internally to the `TMap::serialize(...)` call...

vadi2 replied:
True, fixed

SlySven commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -722,22 +722,23 @@ QString dlgConnectionProfiles::readProfileData(const QString& profile, const QSt
         ofs << what;
-        file.close();
+        if (!file.commit()) {
+            qDebug() << "dlgConnectionProfiles::writeProfileData: error writing profile data: " << file.errorString();
```
Let's report what didn't get written and for which profile:
```suggestion
            qDebug().noquote().nospace() << "dlgConnectionProfiles::writeProfileData(...) ERROR - writing profile: \"" << profile << "\", item: \"" << item << "\", reason: \"" << file.errorString() << "\".";
```

SlySven commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -722,22 +722,23 @@ QString dlgConnectionProfiles::readProfileData(const QString& profile, const QSt
     }
 
-    if (file.error() == QFile::NoError) {
+    if (file.error() == QFileDevice::NoError) {
```
```diff
-    if (file.error() == QFileDevice::NoError) {
-        return qMakePair(true, QString());
-    }
-
-    return qMakePair(false, file.errorString());
+    return (file.error() == QFileDevice::NoError)
+          ? qMakePair(true, QString()))
+          : qMakePair(false, file.errorString());
```

:grinning:

SlySven commented on src/mudlet.cpp:
```diff
@@ -2493,19 +2495,20 @@ QString mudlet::readProfileData(const QString& profile, const QString& item)
         QDataStream ofs(&file);
         ofs << what;
-        file.close();
+        if (!file.commit()) {
+            qDebug() << "dmudlet::writeProfileData: error saving profile data: " << file.errorString();
+        }
```
Same as the `dlgConnectionProfiles` class case, let's report what didn't get written and for which profile:
```suggestion
        if (!file.commit()) {
            qDebug().noquote().nospace() << "mudlet::writeProfileData(...) ERROR - writing profile: \"" << profile << "\", item: \"" << item << "\", reason: \"" << file.errorString() << "\".";
        }
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -4257,17 +4267,17 @@ Hunhandle* mudlet::prepareSharedDictionary()
 #if defined(Q_OS_WIN32)
-    mudlet::self()->sanitizeUtf8Path(affixPathFileName, qsl("profile.dic"));
-    mudlet::self()->sanitizeUtf8Path(dictionaryPathFileName, qsl("profile.aff"));
+    mudlet::self()->sanitizeUtf8Path(affixPath, qsl("profile.dic"));
```
:bulb: Remember this will mess with #6740 ...!

vadi2 committed:
Apply suggestions from code review

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Update src/dlgConnectionProfiles.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Apply review feedback

vadi2 commented:
Hopefully that's everything.

SlySven reviewed (commented):
I seem to have really kicked a hornet's nest with trying to use the ternary operator with two alternate return values around line 738 in `./src/dlgConnectionProfiles.cpp` don't I!

😖

vadi2 commented:
Yeah... I'm not seeing the benefits of shuffling code around like this.

vadi2 committed:
Fix compilation

SlySven commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -734,9 +734,11 @@ QPair<bool, QString> dlgConnectionProfiles::writeProfileData(const QString& prof
+        return {true, QString()};
+    } else {
+        return {false, file.errorString()};
+    };
```
```suggestion
    }
```
😉

SlySven reviewed (approved):
Okay, just one more thing to keep codefactor happy! 😀

vadi2 committed:
Update src/dlgConnectionProfiles.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
That first check (1 of 10) won't succeed until the `development`branch code is merged in as that brings in the `./CI/fix.grid.ui.ordering.js` file that it is currently failing to find - so I'll pull it in now...

SlySven committed:
Merge branch 'development' into use-qsavefile

SlySven committed:
Standarize UI files

SlySven commented:
WTF!!! Where did https://github.com/Mudlet/Mudlet/pull/6738/commits/ebd9b9c5e2064501caad0b01f4df2d7a262b9c65 come from?

😖 

Oh, **golly** it was #6775 wasn't it - ***it wasn't me, Gov, honest!***

SlySven committed:
Merge branch 'development' into use-qsavefile

vadi2 committed:
Merge branch 'development' into use-qsavefile
